### ISSUE TITLE
Getting a request's body seems overly complex

### ISSUE THREAD

**Issue by icambridge** [reactphp/http#7]
So just been in chat asking about getting the request body. I just think it could be improved and this seems like a better forum, since it's got longer text.

@igorw and @cboden pointed out that it's meant to be streamed/piped. And were helpful in pointing me in the direction of the information to get the request body. I can achieve getting the body data in multiple ways, but in each case it seems that I'm getting the data after the request has finished. This could just be me doing it wrong. My latest trial can be found at https://gist.github.com/icambridge/d8c1ea63cce6733bc5d5 I did also manage to get the data using the on data event and again this was executed after the request was over.

I think personally that the body should be an attribute in request object as it clearly belongs to that request. I also think it should be easy to fetch. I do understand the desire to keep memory low when you don't know what size the body will be, however in some cases you have a rough idea and that it could be small. So my idea is to refactor the server class so you can change RequestHeaderParser if you desire to so it returns a BodiedRequest object so if you want it to be a simple getBody you can do so. I'm willing to do the leg work I just don't want to spend time doing then have some point out all the things wrong with that approach.

**Comment by naroga** [comment]
I'm not sure if I understand this completely.
It seems that I have to end() my $response before I can process the data on my request? Is that it?

How do I make a request like this:

    curl -X POST http://myserver -d "name=MyName"

respond with:

    Hello MyName!

?

@cboden @icambridge

**Comment by icambridge** [comment]
@naroga No I believe you can process it without ending the request, once you end the request as far as I understand you can't send any more data.

**Comment by naroga** [comment]
@icambridge please forgive me for not understanding, but can you please provide a simple snippet that retrieves the content payload (the raw request body) without ending the request?

**Comment by naroga** [comment]
If I try the following provided snippet:

    BufferedSink::createPromise($request)->then(function ($body) {
        $response->writeHead(...);
        $response->end("Hello . " $body);
    });

And try the following request

    curl -X POST http://myhost -d 'icambridge'

It will not respond with `Hello icambridge` as excpected. It will just hold the request until curl times out.

**Comment by icambridge** [comment]
```php
BufferedSink::createPromise($request)->then(function ($body) use ($response) {
    $response->write("Hello " . $body);
});

$response->writeHead(....);
```

Should work for you.

**Comment by naroga** [comment]
@icambridge nope. This snippet times out all my requests. If don't explicitly `end()` the `$response`, it doesn't print anything and the browser times out after a minute or so.

**Comment by icambridge** [comment]
So use end at some point?

**Comment by naroga** [comment]
BufferedSink::createPromise($request)->then(function ($body) use ($response) {
    $response->write("Hello " . $body);
    //IF I PUT END HERE, IT NEVER ENDS.
});

$response->writeHead(....);
//IF I PUT END HERE, IT NEVER WRITES 'HELLO $BODY'

**Comment by naroga** [comment]
This is no longer an issue, since pull request #13 has been merged (b36fb250ba0db01e8df752be45a1cae3f1f4b39a).

Sorry about the repeated comments, but this discussion was kind of scattered all around, and I figured it'd be best to comment everywhere just in case future people have the same issues or don't know how to proceed.
